File: src/com/sjsu/cs146/labs/lab7/lab7.java
package com.sjsu.cs146.labs.lab7;
import java.util.*;


public class lab7 {
    public static void main(String[] args) {

    }

    public int findTheCity(int n, int[][] edges, int distanceThreshold) {
        int[][] graph = new int[n][n];
        int[] distances = new int[n];
        boolean[] visited = new boolean[n];

        for (var e : graph) Arrays.fill(e, 1 << 30);

        for (var e : edges) {
            int a = e[0];
            int b = e[1];
            int c = e[2];
            graph[a][b] = c;
            graph[b][a] = c;
        }
        int ans = n;
        int cnt = 1 << 30;
        for (int i = n - 1; i >= 0; i--) {
            int t = dijkstra(i, distances, n, visited, distanceThreshold, graph);
            if (t < cnt) {
                cnt = t;
                ans = i;
            }
        }
        return ans;
    }

    private int dijkstra(int u, int[] distances, int n, boolean[] visited, int distanceThreshold, int[][] graph) {
        Arrays.fill(distances, 1 << 30);
        Arrays.fill(visited, false);
        distances[u] = 0;

        for (int i = 0; i < n; ++i) {
            int k = -1;
            for (int j = 0; j < n; ++j)  if (!visited[j] && (k == -1 || distances[k] > distances[j])) k = j;
            visited[k] = true;
            for (int j = 0; j < n; ++j) distances[j] = Math.min(distances[j], distances[k] + graph[k][j]);
        }

        int cnt = 0;
        for (int d : distances) if (d <= distanceThreshold) ++cnt;
        return cnt;
    }
}

File: src/com/sjsu/cs146/labs/lab7/lab7.py
from typing import List

class Solution:
    def findTheCity(
        self, n: int, edges: List[List[int]], distanceThreshold: int) -> int:

        def dijkstra(u: int) -> int:
            distances = [float('inf')] * n
            distances[u] = 0
            visited = [False] * n
            for _ in range(n):
                k = -1
                for j in range(n): 
                    if not visited[j] and (k == -1 or distances[k] > distances[j]): k = j
                visited[k] = True
                for j in range(n):
                    if distances[k] + g[k][j] < distances[j]: distances[j] = distances[k] + g[k][j]
            return sum(d <= distanceThreshold for d in distances)

        g = [[float('inf')] * n for _ in range(n)]
        for a, b, c in edges: g[a][b] = g[b][a] = c
        ans, cnt = n, float('inf')

        for i in range(n - 1, -1, -1):
            if (t := dijkstra(i)) < cnt: cnt, ans = t, i
        return ans